pkg/otlp/collector.go

```diff
@@ -222,14 +225,26 @@ func BuildAndStart(ctx context.Context, cfg config.Config, s serializer.MetricSe
 	return p, nil
 }
 
-func NewPipelineFromAgentConfig(cfg config.Config, s serializer.MetricSerializer) (*Pipeline, error) {
+func NewPipelineFromAgentConfig(cfg config.Config, s serializer.MetricSerializer, logsAgentChannel chan *message.Message, serverless bool) (*Pipeline, error) {
```

Comment:
One option that would make this easier is to rely on build tags to determine if we're in serverless or not. The `serverless` build tag is used if and only if we are running in serverless mode.

Since all you're using this serverless bool for is for error handling, you could create a new function that changes based on build tags. For example, for a serverless build tag

```go
//go:build otlp && serverless

func validateCfg(cfg config.Config, pcfg PipelineConfig, logsAgentChannel chan *message.Message) error {
    if HasLogsSection(cfg) {
		pipelineError.Store(fmt.Errorf("Cannot enable OTLP log ingestion for serverless"))
		return pipelineError.Load()
	}
	return nil
}
```

while for non-serverless build tags

```go
//go:build otlp && !serverless

func validateCfg(cfg config.Config, pcfg PipelineConfig, logsAgentChannel chan *message.Message) error {
	if pcfg.LogsEnabled && logsAgentChannel == nil {
IbraheemA marked this conversation as resolved.
		pipelineError.Store(fmt.Errorf("OTLP logs is enabled but logs agent is not enabled"))
		return pipelineError.Load()
	}
	return nil
```

Then, you would call this function from here

```go
//go:build otlp

func NewPipelineFromAgentConfig(cfg config.Config, s serializer.MetricSerializer, logsAgentChannel chan *message.Message, serverless bool) (*Pipeline, error) {
	pcfg, err := FromAgentConfig(cfg)
	if err != nil {
		pipelineError.Store(fmt.Errorf("config error: %w", err))
		return nil, pipelineError.Load()
	}
	if err := validateCfg(cfg, pcfg, logsAgentChannel); err != nil {
		return nil, err
	}

	...
}
```

Along those lines or something similar. This is important to us in serverless because we are very concerned about the size of our binary which directly impacts [cold start time](https://www.datadoghq.com/blog/serverless-cold-start-traces/). OTLP support especially adds a huge amount to our binary size already and so I am wary to add any more. Using build tags like this is what we've found to be the best way to keep our binary size down.